signer/src/wsts_state_machine.rs

```diff
@@ -30,33 +39,435 @@ use wsts::traits::Signer as _;
+        signer_public_keys: impl IntoIterator<Item = PublicKey> + Send,
+        threshold: u16,
+        signer_private_key: PrivateKey,
+    ) -> Result<Self, error::Error>
+    where
+        S: storage::DbRead + Send + Sync,
+    {
+        let encrypted_shares = storage
+            .get_encrypted_dkg_shares(aggregate_key)
+            .await?
+            .ok_or(Error::MissingDkgShares(aggregate_key))?;
+
+        let public_dkg_shares: BTreeMap<u32, wsts::net::DkgPublicShares> =
+            BTreeMap::decode(encrypted_shares.public_shares.as_slice())?;
+        let party_polynomials = public_dkg_shares
+            .iter()
+            .flat_map(|(_, share)| share.comms.clone())
+            .collect::<Vec<(u32, PolyCommitment)>>();
+
+        let mut coordinator = Self::new(signer_public_keys, threshold, signer_private_key);
+
+        let aggregate_key = encrypted_shares.aggregate_key.into();
+        coordinator
+            .set_key_and_party_polynomials(aggregate_key, party_polynomials)
+            .map_err(Error::wsts_coordinator)?;
+        coordinator.current_dkg_id = 1;
+
+        coordinator
+            .move_to(WstsState::Idle)
+            .map_err(Error::wsts_coordinator)?;
+
+        Ok(coordinator)
+    }
+
+    fn process_message(
+        &mut self,
+        message: &Message,
+    ) -> Result<(Option<Packet>, Option<OperationResult>), Error> {
+        let packet = Packet::from_message(message);
+        self.0
+            .process_message(&packet)
+            .map_err(Error::wsts_coordinator)
+    }
+
+    fn process_inbound_messages(
+        &mut self,
+        messages: &[Message],
+    ) -> Result<(Vec<Packet>, Vec<OperationResult>), Error> {
+        let packets = messages
+            .iter()
+            .map(Packet::from_message)
+            .collect::<Vec<_>>();
+        self.0
+            .process_inbound_messages(&packets)
+            .map_err(Error::wsts_coordinator)
+    }
+
+    fn process_packet(
+        &mut self,
+        packet: &Packet,
+    ) -> Result<(Option<Packet>, Option<OperationResult>), Error> {
+        self.0
+            .process_message(packet)
+            .map_err(Error::wsts_coordinator)
+    }
+
+    fn process_inbound_packets(
+        &mut self,
+        packets: &[Packet],
+    ) -> Result<(Vec<Packet>, Vec<OperationResult>), Error> {
+        self.0
+            .process_inbound_messages(packets)
+            .map_err(Error::wsts_coordinator)
+    }
+
+    fn start_signing_round(
+        &mut self,
+        message: &[u8],
+        signature_type: SignatureType,
+    ) -> Result<Packet, Error> {
+        self.0
+            .start_signing_round(message, signature_type)
+            .map_err(Error::wsts_coordinator)
     }
 }
 
-#[cfg(any(test, feature = "testing"))]
-impl StateMachineId {
-    /// Create a new identifier
-    pub fn new(value: [u8; 32]) -> Self {
-        StateMachineId(value)
+impl WstsCoordinator for FrostCoordinator {
```

Comment:
nit: would it be possible to template this to avoid having two identical impl for frost and fire with only a single change (`frost::Coordinator` vs `fire::Coordinator`)?